src/queries.ts

```diff
@@ -26,44 +27,63 @@ interface AndroidQueryParams extends BaseQueryParams {
   bid_browsers?: string[];
 }
 
+function isDesktopParams(object: any): object is DesktopQueryParams {
+  return 'bid_window' in object;
+}
+
+function isAndroidParams(object: any): object is AndroidQueryParams {
+  return 'bid_android' in object;
+}
+
 // Constructs a query that returns a fully-detailed list of events from the merging of several sources (window, afk, web).
-// Puts it's results in `not_afk` and `events`.
-function canonicalEvents(params: DesktopQueryParams): string {
+// Performs:
+//  - AFK filtering (if filter_afk is true)
+//  - Categorization (if classes specified)
+//  - Filters by category (if filter_classes set)
+// Puts it's results in `events` and `not_afk` (if not_afk available for platform).
+function canonicalEvents(params: DesktopQueryParams | AndroidQueryParams): string {
+  // Needs escaping for regex patterns like '\w' to work (JSON.stringify adds extra unecessary escaping)
+  const classes_str = JSON.stringify(params.classes).replace('\\\\', '\\');
+  const cat_filter_str = JSON.stringify(params.filter_classes);
+  const bid_window = isDesktopParams(params) ? params.bid_window : params.bid_android;
+
   return `
-    events  = flood(query_bucket("${params.bid_window}"));
-    not_afk = flood(query_bucket("${params.bid_afk}"));
-    not_afk = filter_keyvals(not_afk, "status", ["not-afk"]);
-    ${params.filter_afk ? 'events  = filter_period_intersect(events, not_afk);' : ''}
+    events = flood(query_bucket("${bid_window}"));
+    ${
+      isDesktopParams(params)
+        ? `not_afk = flood(query_bucket("${params.bid_afk}"));
+           not_afk = filter_keyvals(not_afk, "status", ["not-afk"]);`
+        : ''
+    }
+    ${isAndroidParams(params) ? 'events = merge_events_by_keys(events, ["app"]);' : ''}
+
+    ${
+      isDesktopParams(params) && params.filter_afk
+        ? 'events = filter_period_intersect(events, not_afk);'
+        : ''
+    }
+    ${params.classes ? `events = categorize(events, ${classes_str});` : ''}
```

Comment:
If we in the future also want to classify browser events we need to move this again, maybe it doesn't make sense to have it in canonicalEvents?

Reply:
The idea is to move browser events into canonical query, so no.